cobc/typeck.c

```diff
@@ -5772,41 +5785,41 @@ expr_expand (cb_tree *x)
 			*x = p->x;
 			goto start;
 		}
-		expr_expand (&p->x);
+		build_expr_expand (&p->x);
 		if (p->y) {
-			expr_expand (&p->y);
+			build_expr_expand (&p->y);
 		}
 	}
 }
 
 static cb_tree
-cb_expr_finish (void)
+build_expr_finish (void)
 {
-	/* Reduce all */
-	(void)expr_reduce (0);
+	/* Reduce all (prio of token 0 is smaller than all other ones)*/
+	(void)build_expr_reduce (0);
+
+	expr_stack[TOPSTACK].value->source_file = cb_source_file;
+	expr_stack[TOPSTACK].value->source_line = cb_exp_line;
 
-	if (!expr_stack[3].value) {
+	if (expr_index != TOPSTACK+1) {
 		/* TODO: Add test case for this to syn_misc.at invalid expression */
-		cb_error (_("invalid expression"));
+		cb_error_x (expr_stack[TOPSTACK].value, _("invalid expression: unfinished expression"));
 		return cb_error_node;
 	}
 
-	expr_stack[3].value->source_file = cb_source_file;
-	expr_stack[3].value->source_line = cb_exp_line;
-
-	if (expr_index != 4) {
+	if (!expr_stack[TOPSTACK].value) {
 		/* TODO: Add test case for this to syn_misc.at invalid expression */
-		cb_error_x (expr_stack[3].value, _("invalid expression"));
+		cb_error (_("invalid expression: final expression is not a value"));
```

Comment:
please add a testcase for that, then drop the TODO note (ideally also for the one below)

Reply:
Hum... I don't really see when such a case can happen, my understanding of this code is still superficial...